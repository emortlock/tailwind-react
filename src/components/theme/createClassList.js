import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
