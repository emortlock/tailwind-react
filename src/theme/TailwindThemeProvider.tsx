import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
