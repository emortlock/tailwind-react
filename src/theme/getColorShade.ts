const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
