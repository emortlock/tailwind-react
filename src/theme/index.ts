export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
