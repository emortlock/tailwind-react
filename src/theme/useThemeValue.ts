import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
