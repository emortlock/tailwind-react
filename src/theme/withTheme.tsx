import type { Theme } from './TailwindTheme'

import React from 'react'

import { TailwindTheme } from './TailwindTheme'

export interface WithTheme {
  theme: Theme
}

export const withTheme = <P extends React.PropsWithChildren<{ theme: Theme }>>(
  Component: React.FC<P>,
) => {
  const WithTheme: React.FC<Omit<P, 'theme'>> = (props) => (
    <TailwindTheme.Consumer>
      {(theme) => {
        // @ts-ignore FIXME: Constraint error
        return <Component {...props} theme={theme} />
      }}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}
