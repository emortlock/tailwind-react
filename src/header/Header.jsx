import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'
import { getUniqueID } from '../utils'
import { Container } from '../container'

class Header extends PureComponent {
  constructor(props) {
    super(props)

    this.state = {
      open: false,
      collapsable: true,
    }

    this.mql = null
    this.id = getUniqueID('header')

    this.handleMatch = this.handleMatch.bind(this)
    this.handleToggle = this.handleToggle.bind(this)
  }

  componentDidMount() {
    const { theme } = this.props
    if (window.matchMedia) {
      this.mql = window.matchMedia(`(min-width: ${theme.breakpoints.lg})`)
      this.mql.addListener(this.handleMatch)

      if (this.mql.matches) {
        this.handleMatch(this.mql)
      }
    }
  }

  componentWillUnmount() {
    if (this.mql) this.mql.removeListener(this.handleMatch)
  }

  handleToggle(forceState) {
    const { open } = this.state

    this.setState({
      open: forceState || !open,
    })
  }

  handleMatch(mql) {
    this.setState({ collapsable: !mql.matches })
  }

  render() {
    const { open, collapsable } = this.state
    const { theme, is, children, bg, text, ...rest } = this.props

    const headerProps = {
      id: this.id,
      style: {
        bg: bg || theme.brandColors.primary,
        text: text || theme.textColors.on.primary,
      },
      open,
      collapsable,
      onToggle: this.handleToggle,
    }

    return (
      <BaseComponent
        is={is}
        id={headerProps.id}
        bg={headerProps.style.bg}
        text={headerProps.style.text}
        p={{ y: theme.spacing.md }}
        role="banner"
        {...rest}
      >
        <Container
          flex={[true, 'wrap']}
          items="center"
          justify="between"
          padding
        >
          {React.Children.map(children, child =>
            React.cloneElement(child, { header: headerProps }),
          )}
        </Container>
      </BaseComponent>
    )
  }
}

Header.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  bg: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
}

Header.defaultProps = {
  is: 'header',
  children: undefined,
  bg: undefined,
  text: undefined,
}

export { Header as component }
export default withTheme(Header)
