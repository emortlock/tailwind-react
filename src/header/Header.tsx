import { BoxProps } from 'tailwind-react-primitives'
import { WithTheme, Theme } from '../theme'

import React from 'react'

import { RawBox as Box, Flex, filterProps } from 'tailwind-react-primitives'
import { withTheme } from '../theme'
import { getUniqueID } from '../utils'
import { RawContainer as Container } from '../container'

export interface WithHeader {
  header?: {
    id?: string
    style?: {
      bg?: string | string[]
      text?: string | string[]
    }
    isOpen?: boolean
    isCollapsable?: boolean
    onToggle?: (forceState?: boolean) => void
    screen?: string
  }
}

export interface HeaderProps<E extends HTMLElement = HTMLDivElement>
  extends WithTheme,
    BoxProps<E> {
  screen?: keyof Theme['breakpoints']
}

export const RawHeader = <E extends HTMLElement = HTMLDivElement>({
  theme,
  id,
  screen = 'lg',
  as = 'header',
  children,
  bg,
  text,
  ...rest
}: HeaderProps<E>) => {
  const [isOpen, setOpen] = React.useState(false)
  const [isCollapsable, setCollapsable] = React.useState(false)

  const [headerId, setHeaderId] = React.useState(id)
  React.useEffect(() => {
    if (headerId) return

    setHeaderId(getUniqueID('neader'))
  })

  const mql = React.useRef<MediaQueryList | undefined>()
  const handleMatch = React.useCallback(
    (mql: MediaQueryList | MediaQueryListEvent) => {
      setCollapsable(!mql.matches)
    },
    [],
  )
  React.useEffect(() => {
    if (screen && window.matchMedia) {
      mql.current = window.matchMedia(
        `(min-width: ${theme.breakpoints[screen]})`,
      )
      mql.current.addEventListener('change', handleMatch)

      if (mql.current.matches) {
        handleMatch(mql.current)
      }
    }

    return () => {
      if (mql.current) mql.current.removeEventListener('change', handleMatch)
    }
  }, [])

  const handleToggle = React.useCallback((forceState?: boolean) => {
    setOpen(forceState ?? ((currentState) => !currentState))
  }, [])

  const bgStyle = bg || theme.brandColors.primary
  const textStyle = text || theme.textColors.on.primary

  const headerProps: WithHeader['header'] = {
    id: headerId ?? '',
    style: {
      bg: bgStyle,
      text: textStyle,
    },
    isOpen,
    isCollapsable,
    onToggle: handleToggle,
    screen,
  }

  return (
    <Box<E>
      as={as}
      id={headerId}
      bg={bgStyle}
      text={textStyle}
      p={{ y: theme.spacing.md }}
      role="banner"
      {...filterProps(rest, ['id'])}
    >
      <Container<HTMLDivElement>
        theme={theme}
        as={(props) => <Flex {...props} wrap />}
        items="center"
        justify="between"
        padding
      >
        {React.Children.map(children, (child) =>
          React.isValidElement(child)
            ? React.cloneElement<WithHeader>(child, { header: headerProps })
            : child,
        )}
      </Container>
    </Box>
  )
}

export const Header = withTheme(RawHeader)
