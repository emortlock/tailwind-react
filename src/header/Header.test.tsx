import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawHeader as Header } from './Header'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign({ theme: defaultTheme }, testProps)

  const wrapper = render(<Header {...props}>Content</Header>)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('Header', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup({ bg: 'blue-400', text: 'white' })

    expect(asFragment()).toMatchSnapshot()
  })
})
