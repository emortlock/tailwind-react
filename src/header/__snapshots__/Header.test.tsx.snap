// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders matching snapshot 1`] = `
<DocumentFragment>
  <header
    class="bg-blue-400 text-white py-4"
    id="neader-1"
    role="banner"
  >
    <div
      class="flex flex-wrap mx-auto px-4 container items-center justify-between"
    >
      Content
    </div>
  </header>
</DocumentFragment>
`;
